Commited ZDC Module it contains the first version of ZDCRawDataTask. …

cpuggion84 opened the pull request:
Created new module ZDC.
Created new task ZDCRawDataTask
Created new checker ZDCRawDataCheck (Created but not yet implemented)
Inside task ZDCRawDataTask produced these Histograms:
26 histograms for pedestals: will be used to provide calibration object for reconstruction (backup average pedestal information,
in case of missing information in current event and current orbit)
1 pedestal summary for central shifter
26 scalers: can be used to produce graphs of luminosity evolution (TBD)
26  signals “oscilloscope-like” waveforms
10 BC maps to align ALICE trigger
10 BC maps to align AUTO trigger
10 BC maps to align AUTO trigger or ALICE trigger
1 channels trasmitted
1 channels fired

knopers8 commented on Modules/ZDC/include/ZDC/ZDCRawDataCheck.h:
```diff
@@ -0,0 +1,46 @@
+
+///
+/// \file   ZDCRawDataCheck.h
+/// \author My Name
```
Please put your name here and below

cpuggion84 replied:
Ok I'll add these information in the next commit.

knopers8 commented on Modules/ZDC/include/ZDC/ZDCRawDataCheck.h:
```diff
@@ -0,0 +1,46 @@
+{
+
+/// \brief  Example QC Check
+/// \author My Name
```
Please fix the doxygen

knopers8 commented on Modules/ZDC/src/ZDCRawDataCheck.cxx:
```diff
@@ -0,0 +1,85 @@
+
+///
+/// \file   ZDCRawDataCheck.cxx
+/// \author My Name
```
here as well

cpuggion84 replied:
Ok I'll add these information in the next commit.

knopers8 commented on Modules/ZDC/src/ZDCRawDataCheck.cxx:
```diff
@@ -0,0 +1,85 @@
+
+void ZDCRawDataCheck::configure() {}
+
+Quality ZDCRawDataCheck::check(std::map<std::string, std::shared_ptr<MonitorObject>>* moMap)
```
If this check is not actually doing anything meaningful, I would recommend to remove it.

cpuggion84 replied:
I did not yet implemented the check of histograms. I created the checker for raw histograms because we had some histograms that we want use for the shifter. I wanted try if the json file configuration was correct and if when run the task runned also the checker.  If you are agree i prefer don't remove it,because in the next commites i'll implement it.

knopers8 replied:
Sure, no problem. Just FYI, if you removed the check from the json, it should still run the task and store the objects.

knopers8 commented on Modules/ZDC/src/ZDCRawDataTask.cxx:
```diff
@@ -0,0 +1,1032 @@
+{
+  ILOG(Info, Support) << "initialize ZDCRawDataTask" << ENDM; // QcInfoLogger is used. FairMQ logs will go to there as well.
+
+  // this is how to get access to custom parameters defined in the config file at qc.tasks.<task_name>.taskParameters
+  if (auto param = mCustomParameters.find("myOwnKey"); param != mCustomParameters.end()) {
+    ILOG(Info, Devel) << "Custom parameter - myOwnKey: " << param->second << ENDM;
+  }
```
no need to keep this
```suggestion
```

cpuggion84 replied:
Ok I'll remove it.

knopers8 commented on Modules/ZDC/src/ZDCRawDataTask.cxx:
```diff
@@ -0,0 +1,1032 @@
+  init();
+  mHistogram = new TH1F("eventSize", "eventSize", 20, 0, 30000);
+  getObjectsManager()->startPublishing(mHistogram);
+  getObjectsManager()->startPublishing(new TH1F("example2", "example2", 20, 0, 30000));
+  try {
+    getObjectsManager()->addMetadata(mHistogram->GetName(), "custom", "34");
+  } catch (...) {
+    // some methods can throw exceptions, it is indicated in their doxygen.
+    // In case it is recoverable, it is recommended to catch them and do something meaningful.
+    // Here we don't care that the metadata was not added and just log the event.
+    ILOG(Warning, Support) << "Metadata could not be added to " << mHistogram->GetName() << ENDM;
+  }
```
this is probably some example that might not be needed in the detector QC
```suggestion
```

cpuggion84 replied:
Ok I'll remove it.

knopers8 commented on Modules/ZDC/src/ZDCRawDataTask.cxx:
```diff
@@ -0,0 +1,1032 @@
+
+                if (ih < (int)fMatrixHistoBaseline[mod][ch].size()){
+                    getObjectsManager()->startPublishing(fMatrixHistoBaseline[mod][ch].at(ih).histo);
+                    try {
+                        getObjectsManager()->addMetadata(fMatrixHistoBaseline[mod][ch].at(ih).histo->GetName(), "custom", "34");
+                        return true;
+                    }
+                    catch (...) {
+                        ILOG(Warning, Support) << "Metadata could not be added to " << fMatrixHistoBaseline[mod][ch].at(ih).histo->GetName() << ENDM;
+                        return false;
+                    }
```
you probably do not need this, it was an example
```suggestion
```

cpuggion84 replied:
Ok I'll remove it.

knopers8 commented on Modules/ZDC/src/ZDCRawDataTask.cxx:
```diff
@@ -0,0 +1,1032 @@
+
+void ZDCRawDataTask::reset()
+{
+  // clean all the monitor objects here
+
+  ILOG(Info, Support) << "Resetting the histogram" << ENDM;
+  mHistogram->Reset();
```
Please reset all the monitor objects here. If you don't, several functionalities will not work - e.g. merging, moving windows, start->stop->start sequences.

cpuggion84 replied:
Ok I'll do it

knopers8 commented on Modules/ZDC/src/ZDCRawDataTask.cxx:
```diff
@@ -0,0 +1,1032 @@
+
+                if (ih < (int)fMatrixHistoCounts[mod][ch].size()){
+                    getObjectsManager()->startPublishing(fMatrixHistoCounts[mod][ch].at(ih).histo);
+                    try {
+                        getObjectsManager()->addMetadata(fMatrixHistoCounts[mod][ch].at(ih).histo->GetName(), "custom", "34");
+                        return true;
+                    }
+                    catch (...) {
+                        ILOG(Warning, Support) << "Metadata could not be added to " << fMatrixHistoCounts[mod][ch].at(ih).histo->GetName() << ENDM;
+                        return false;
+                    }
```
```suggestion
```
you probably don't need this, it was an example

cpuggion84 replied:
Ok i'll remove it.

knopers8 commented on Modules/ZDC/src/ZDCRawDataTask.cxx:
```diff
@@ -0,0 +1,1032 @@
+
+                if (ih < (int)fMatrixHistoSignal[mod][ch].size()){
+                    getObjectsManager()->startPublishing(fMatrixHistoSignal[mod][ch].at(ih).histo);
+                    try {
+                        getObjectsManager()->addMetadata(fMatrixHistoSignal[mod][ch].at(ih).histo->GetName(), "custom", "34");
+                        return true;
+                    }
+                    catch (...) {
+                        ILOG(Warning, Support) << "Metadata could not be added to " << fMatrixHistoSignal[mod][ch].at(ih).histo->GetName() << ENDM;
+                        return false;
+                    }
```
```suggestion
```
same here and below

knopers8 commented on Modules/ZDC/src/ZDCRawDataTask.cxx:
```diff
@@ -0,0 +1,1032 @@
+    ILOG(Info, Support) << TString::Format("Key Word %s",TokenString.at(0).c_str()) << ENDM;
+
+    if (strcmp(TokenString.at(0).c_str(), "MODULE") == 0) {
+        if (decodeModule(TokenString,LineNumber) == false) return false;
+        else return true;
```
```suggestion
        return decodeModule(TokenString,LineNumber);
```
can't all of those be simplified to something like this?

cpuggion84 replied:
Yes I can change it

knopers8 commented on Modules/ZDC/src/ZDCRawDataTask.cxx:
```diff
@@ -0,0 +1,1032 @@
+
+bool ZDCRawDataTask::decodeBaseline(std::vector<std::string> TokenString,int LineNumber){
+   if (TokenString.size() == 5){
+      if (addNewHisto(TokenString.at(0),TokenString.at(1),TokenString.at(2),TokenString.at(3),TokenString.at(4))== true) return true;
+      else return false;
```
```suggestion
      return addNewHisto(TokenString.at(0),TokenString.at(1),TokenString.at(2),TokenString.at(3),TokenString.at(4));
```
same here and in other methods below

cpuggion84 replied:
Yes I can change it

knopers8 reviewed (approved):
Thank you @cpuggion84 . I approve the PR to let the CI run (you are a first-time contributor, so it will not run without a positive review), but please consider addressing my comments.

knopers8 commented:
I think you are also missing an `add_subdirectory` statement in Modules/CMakeLists.txt for your module

cpuggion84 commented:
I think you are also missing an add_subdirectory statement in Modules/CMakeLists.txt for your module
Sorry is not clear for me how  to fix it. Can You give me some instruction?

I see that there are Clang format errors. How can check it, before to do the commit?

Thank you in advanced

knopers8 commented:
> Sorry is not clear for me how to fix it. Can You give me some instruction?

Sure. You just need to add this line in `Modules/CMakeLists.txt`, otherwise your new module will not be build at all.
```
add_subdirectory(ZDC)
```
Actually, it should have been added by the `o2-qc-module-configurator.sh` script, so I am a bit surprised it is missing in your case.

> I see that there are Clang format errors. How can check it, before to do the commit?

Here are instruction how to format your code:
https://github.com/AliceO2Group/CodingGuidelines#formatting-tool

knopers8 reviewed (changes requested):
blocking merging, so it is not accidentally merged until the module is actually added to Modules/CMakeLists.txt

cpuggion84 commented:
Dear Piotr,
Yesterday i tried to commit the changes that you asked me. I committed before the file CMakeLists.txt and after the others updates that you asked me. But involuntary with git I created a new pull request and i see only the changes of file CMakeLists.txt . How can merge all files in my first pull request? sorry but for me it is the first time that use git in this way.

knopers8 commented:
You should be able to apply it like this, unless your repository is in some weird state:
```
# first go to your repository
git checkout feature-new-stuff # to select the branch for this PR
vim Modules/CMakeLists.txt # modify the file with your favourite editor
git reset # to make sure you did not already cache other changes to the upcoming commit
git add Modules/CMakeLists.txt # this adds the modified file to the upcoming commit 
git commit # name the commit and save it
git push origin feature-new-stuff # pushes the new commit to the feature-new-stuff branch on your fork. it will appear in this PR
```

cpuggion84 commented:
During this new committed i fixed the issue regarding:
 - Filled doxygen
 - Modules/CMakeLists.txt
 - Clang format fixed
 - Simplified the code where request
 -  Change the json configuration file, so now don't start the checker.
 - addMetadata I can't remove it, because crashed the task. I put the description of histogram.

knopers8 commented:
Please fix the legit warning from the mac build:
```
/System/Volumes/Data/build/ci_checks/alice_ci_0/qualitycontrol-o2/sw/SOURCES/QualityControl/1046/0/Modules/ZDC/include/ZDC/ZDCRawDataCheck.h:37:8: error: 'configure' overrides a member function but is not marked 'override' [-Werror,-Winconsistent-missing-override]
  void configure();
       ^
 ```
Then we are good to go
